File: 剑指offer-2012年书籍版/Java/No12.java
/**
 * 输入数字n，按顺序打印出从1最大的n位十进制数。
 * 比如输入3，则打印出1、2、3一直到最大的3位数即999
 */

public class No12 {

    public static void main(String[] args) {
        printNum(3);
    }

    private static void printNum(int n) {
        if (n < 0)
            return;
        int[] array = new int[n];
        printArray(array, 0);
    }

    private static void printArray(int[] array, int n) {

        if (n != array.length) {
            for (int i = 0; i < 10; i++) {
                array[n] = i;
                printArray(array, n + 1);
            }
        } else {
            boolean flag = false;
            for (int j = 0; j < array.length; j++) {
                if (array[j] != 0) {
                    flag = true;
                }
                if (flag) {
                    System.out.print(array[j]);
                }
            }
            // 去掉空白行
            if (flag) {
                System.out.println();
            }
        }
    }
}

File: 剑指offer-2012年书籍版/Kotlin/No12.kt
/**
 * 输入数字n，按顺序打印出从1最大的n位十进制数。
 * 比如输入3，则打印出1、2、3一直到最大的3位数即999
 */

class No12 {

    fun main() {
        printNum(4)
    }

    fun printNum(n: Int) {
        if (n <= 0) {
            return
        }
        printArray(Array<Int>(n) { 0 }, 0)
    }

    fun printArray(array: Array<Int>, index: Int) {
        when (index) {
            array.size -> {
                val numStr = array.joinToString(separator = "") { it.toString() }
                val nonZeroIndex = numStr.indexOfFirst { it != '0' }
                if (nonZeroIndex == -1) {
                    println('0')
                } else {
                    println(numStr.substring(nonZeroIndex))
                }
            }
            else -> {
                for (i in 0..9) {
                    array[index] = i
                    printArray(array, index + 1)
                }
            }
        }
    }
}